### ISSUE TITLE
os/exec: TestExtraFilesRace flake on plan9 since 2022-11-16

### ISSUE THREAD

**Issue by millerresearch** [golang/go#57180]
#!watchflakes
    post <- pkg == "os/exec" && test == "TestExtraFilesRace" && goos == "plan9"

On the plan9-arm builder the os/exec test sometimes fails like this:
```
--- FAIL: TestExtraFilesRace (0.50s)
    exec_test.go:850: iteration 2, process B got:
        
        want:
        fd3: listener 127.0.0.1:40881
        
FAIL
FAIL	os/exec	1.262s
```
This has been seen since 2022-11-16 (earliest log [here](https://build.golang.org/log/c92debe49346e0156c9b8aecedf5ba9997dd0b81)).

The last change to TextExtraFilesRace before that date was the addition of `t.Parallel()` in [CL 439196](https://go-review.googlesource.com/c/go/+/439196
). The comment for that change says

> Tests that use Setenv or Chdir or check for FDs opened during the test
still cannot be parallelized,

@bcmills, I haven't looked deeply into this but it does seem a FD is being opened during the test - is this a potential trouble spot?

I have managed to replicate the error by running `go tool dist test go_test:net/url go_test:os go_test:os/exec` enough times. If I remove the `t.Parallel()`, the error seems not to occur.

**Comment by bcmills** [comment]
`greplogs -l -e '\Aplan9.*(\n.*)*FAIL: TestExtraFilesRace' --since=2022-01-01`
[2022-12-05T22:01:50-185e1a7/plan9-arm](https://build.golang.org/log/07ea04f0a7736675739bb60d17a29d102498c405)
[2022-12-05T16:41:48-ad55b87/plan9-arm](https://build.golang.org/log/dd058eed68e1e46794c6f2fd52172b2d8af5de2f)
[2022-12-02T23:40:37-c0497d1/plan9-arm](https://build.golang.org/log/498d0b9fc2a85a13b45b27887c954ffdc26ba5f2)
[2022-12-01T21:00:47-cd133ab/plan9-arm](https://build.golang.org/log/e9494db6b65487a70a76b05a5293f016dfde48bf)
[2022-11-28T16:06:53-76ec47e/plan9-arm](https://build.golang.org/log/dc3f3559303045554257f28b4e5f61de6d80b72b)
[2022-11-24T03:05:33-7a9ce74/plan9-arm](https://build.golang.org/log/3760f08c5fc9b6975ba954e22d96f0be9ff63d2e)
[2022-11-23T19:34:12-a8510f9/plan9-arm](https://build.golang.org/log/a171413c80a2721ea8e78dfdfaacfe1c6f0b98f1)
[2022-11-22T18:52:58-21015cf/plan9-arm](https://build.golang.org/log/8ffd11f592c028db00b1d1c4db448bf01e876275)
[2022-11-21T22:16:06-995c0f3/plan9-arm](https://build.golang.org/log/fbc099d62dc896215ed98678cf2970baabb8d688)
[2022-11-21T16:19:15-d7812ab/plan9-arm](https://build.golang.org/log/1dd9a0233fdab900235174ec23422a64115f4049)
[2022-11-21T15:06:26-8614c52/plan9-arm](https://build.golang.org/log/9e5450945d9a029b5489a8925a1e04ec074d5ec3)
[2022-11-18T22:32:44-c824448/plan9-arm](https://build.golang.org/log/ac40319b85c622a1683b090d02c45706ffe54d58)
[2022-11-18T21:12:24-dccc58e/plan9-arm](https://build.golang.org/log/2dc93a4dead76808bddb232ef8e79b4c8a901116)
[2022-11-18T17:59:44-b2faff1/plan9-arm](https://build.golang.org/log/88d02c2ed8b385af69052ece2adfc66bdf386e14)
[2022-11-18T14:08:47-349d398/plan9-arm](https://build.golang.org/log/99c12fc435f2c74b3d812d99f6bf00bb41c7e891)
[2022-11-17T18:53:13-5f7abec/plan9-arm](https://build.golang.org/log/a406efc0aac0446fdaffa564fee1c5cb9f73ba75)
[2022-11-17T18:24:36-217ed95/plan9-arm](https://build.golang.org/log/dddab4d1f440ff972d93bbc0450374f85f8e4a01)
[2022-11-16T23:17:58-6d0bf43/plan9-arm](https://build.golang.org/log/90b6f2fa3aaecece641830e7763a3aa80b4cff9f)
[2022-11-16T21:42:58-52d9e41/plan9-arm](https://build.golang.org/log/6c99a06ee901a40bf61158314bcde19fc4df456a)
[2022-11-16T21:35:23-898d000/plan9-arm](https://build.golang.org/log/c92debe49346e0156c9b8aecedf5ba9997dd0b81)
[2022-09-13T17:41:06-6f7c9ea/plan9-arm](https://build.golang.org/log/5a11c5b876c9e4a322f9fc96491657bfc21741e3)

**Comment by bcmills** [comment]
The one on 2022-09-13 has a funky failure mode, which suggests the regression happened on 2022-11-16.

**Comment by bcmills** [comment]
(Note that I added the call to `t.Parallel()` in that test in [CL 439196](https://go.dev/cl/439196), merged 2022-10-10.)

**Comment by millerresearch** [comment]
> The one on 2022-09-13 has a funky failure mode

Errors like "/boot/workdir/... does not exist' are evidence of a flake in the underlying Plan 9 filesystem, not specifically relevant to this issue.

**Comment by millerresearch** [comment]
> for some reason `cmdDescribeFiles` in the subprocess fails to wrap it in a `net.FileListener`.

The error returned from the attempt to wrap it is `file+net fd3: fd2path: fd out of range or not open`

**Comment by bcmills** [comment]
That suggests perhaps a problem in either `syscall.StartProcess` (failing to actually forward the extra FD) or perhaps `(*net.TCPListener).File` (silently failing to produce a valid FD, or producing a duplicate?).

**Comment by millerresearch** [comment]
> That suggests perhaps a problem in either `syscall.StartProcess` (failing to actually forward the extra FD)

You're right, a logic error was sometimes causing the newly dup'ed FD to be closed.

**Comment by bcmills** [comment]
> suggests the regression happened on or about 2022-11-16.

That turned out to be off by 10½ years. 😅

(Based on the fix CL, it appears that FD forwarding on plan9 has been subtly broken ~forever, but the specific bug here was part of an attempted fix for a related bug in [CL 6009046](https://go.dev/cl/6009046).)

**Comment by bcmills** [comment]
The above failures predate the fix (see #57632).
